File: HSE/ADS_contests/Graphs_2/NlogM_dijkstra.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <limits>
#include <utility>

using namespace std;

const long long INF = 2009000999;

vector<long long> dijkstra(const vector<vector<pair<int, int>>>& graph, int start, int n) {
    vector<long long> distances(n, INF);
    distances[start] = 0;
    priority_queue<pair<long long, int>, vector<pair<long long, int>>, greater<pair<long long, int>>> pq;
    pq.push({0, start});
    
    vector<bool> visited(n, false);
    
    while (!pq.empty()) {
        int current_vertex = pq.top().second;
        long long current_distance = pq.top().first;
        pq.pop();
        
        if (visited[current_vertex]) {
            continue;
        }
        
        visited[current_vertex] = true;
        
        for (const auto& edge : graph[current_vertex]) {
            int neighbor = edge.first;
            int weight = edge.second;
            
            if (visited[neighbor]) {
                continue;
            }
            
            long long distance = current_distance + weight;
            
            if (distance < distances[neighbor]) {
                distances[neighbor] = distance;
                pq.push({distance, neighbor});
            }
        }
    }
    
    return distances;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    
    int num_tests;
    cin >> num_tests;
    
    for (int test = 0; test < num_tests; ++test) {
        int n, m;
        cin >> n >> m;
        
        // Initialize graph as adjacency list
        vector<vector<pair<int, int>>> graph(n);
        
        for (int i = 0; i < m; ++i) {
            int u, v, w;
            cin >> u >> v >> w;
            
            graph[u].push_back({v, w});
            graph[v].push_back({u, w});
        }
        
        // // Process duplicate edges
        // for (int i = 0; i < n; ++i) {
        //     if (graph[i].empty()) continue;
            
        //     // Sort edges by neighbor
        //     sort(graph[i].begin(), graph[i].end());
            
        //     // Remove duplicates keeping minimum weight
        //     vector<pair<int, int>> unique_edges;
        //     unique_edges.push_back(graph[i][0]);
            
        //     for (size_t j = 1; j < graph[i].size(); ++j) {
        //         if (graph[i][j].first == unique_edges.back().first) {
        //             // Same neighbor, keep minimum weight
        //             unique_edges.back().second = min(unique_edges.back().second, graph[i][j].second);
        //         } else {
        //             unique_edges.push_back(graph[i][j]);
        //         }
        //     }
            
        //     graph[i] = std::move(unique_edges);
        // }
        
        int start;
        cin >> start;
        
        // Run Dijkstra's algorithm
        vector<long long> distances = dijkstra(graph, start, n);
        
        // Print distances
        for (int i = 0; i < n; ++i) {
            cout << distances[i];
            if (i < n - 1) {
                cout << " ";
            }
        }
        cout << endl;
    }
    
    return 0;
}

File: HSE/ADS_contests/Graphs_2/NlogM_dijkstra.py
import heapq

def dijkstra(edges, start, n):
    distances = [2009000999] * n
    distances[start] = 0
    

    pq = [(0, start)]
    visited = [False] * n
    
    while pq:
        current_distance, current_vertex = heapq.heappop(pq)
        
        if visited[current_vertex]:
            continue
        
        visited[current_vertex] = True
        
        for neighbor, weight in edges[current_vertex]:
            if visited[neighbor]:
                continue
                
            distance = current_distance + weight
            
            if distance < distances[neighbor]:
                distances[neighbor] = distance
                heapq.heappush(pq, (distance, neighbor))
    
    return distances

def main():
    num_tests = int(input())
    
    for _ in range(num_tests):
        n, m = map(int, input().split())
        
        edges = [[] for _ in range(n)]
        
        for _ in range(m):
            u, v, w = map(int, input().split())
            
            edges[u].append((v, w))
            edges[v].append((u, w))
        
        for i in range(n):
            neighbor_weights = {}
            for neighbor, weight in edges[i]:
                if neighbor in neighbor_weights:
                    neighbor_weights[neighbor] = min(neighbor_weights[neighbor], weight)
                else:
                    neighbor_weights[neighbor] = weight
            
            edges[i] = [(neighbor, weight) for neighbor, weight in neighbor_weights.items()]
        
        start = int(input())
        
        # Run Dijkstra's algorithm
        distances = dijkstra(edges, start, n)
        
        # Print distances
        print(' '.join(map(str, distances)))

if __name__ == "__main__":
    main()